// vuetify/lib/labs/VNumberInput/VNumberInput.mjs
import { Fragment as _Fragment, mergeProps as _mergeProps, createVNode as _createVNode } from "vue";
// Styles
import "./VNumberInput.css";

// Components
import { VBtn } from "../../components/VBtn/index.mjs";
import { VDefaultsProvider } from "../../components/VDefaultsProvider/index.mjs";
import { VDivider } from "../../components/VDivider/index.mjs";
import { filterFieldProps, makeVFieldProps, VField } from "../../components/VField/VField.mjs";
import { makeVInputProps, VInput } from "../../components/VInput/VInput.mjs"; // Composables
import { makeFocusProps, useFocus } from "../../composables/focus.mjs";
import { useProxiedModel } from "../../composables/proxiedModel.mjs"; // Utilities
import { computed, ref, watchEffect } from 'vue';
import { clamp, filterInputAttrs, genericComponent, getDecimals, only, propsFactory, useRender } from "../../util/index.mjs"; // Types
const makeVNumberInputProps = propsFactory({
  controlVariant: {
    type: String,
    default: 'default'
  },
  inset: Boolean,
  hideInput: Boolean,
  min: {
    type: Number,
    default: -Infinity
  },
  max: {
    type: Number,
    default: Infinity
  },
  step: {
    type: Number,
    default: 1
  },
  ...only(makeVInputProps(), ['density', 'disabled', 'focused', 'hideDetails', 'hint', 'label', 'persistentHint', 'readonly']),
  ...only(makeVFieldProps(), ['baseColor', 'bgColor', 'class', 'color', 'disabled', 'error', 'loading', 'reverse', 'rounded', 'style', 'theme', 'variant']),
  ...makeFocusProps()
}, 'VNumberInput');
export const VNumberInput = genericComponent()({
  name: 'VNumberInput',
  inheritAttrs: false,
  props: {
    ...makeVNumberInputProps(),
    modelValue: {
      type: Number,
      default: undefined
    }
  },
  emits: {
    'update:modelValue': val => true
  },
  setup(props, _ref) {
    let {
      attrs,
      emit,
      slots
    } = _ref;
    const model = useProxiedModel(props, 'modelValue');
    const {
      isFocused,
      focus,
      blur
    } = useFocus(props);
    const inputRef = ref();
    const stepDecimals = computed(() => getDecimals(props.step));
    const modelDecimals = computed(() => model.value != null ? getDecimals(model.value) : 0);
    const canIncrease = computed(() => {
      if (model.value == null) return true;
      return model.value + props.step <= props.max;
    });
    const canDecrease = computed(() => {
      if (model.value == null) return true;
      return model.value - props.step >= props.min;
    });
    watchEffect(() => {
      if (model.value != null && (model.value < props.min || model.value > props.max)) {
        model.value = clamp(model.value, props.min, props.max);
      }
    });
    function onFocus() {
      if (!isFocused.value) focus();
    }
    const controlVariant = computed(() => {
      return props.hideInput ? 'stacked' : props.controlVariant;
    });
    const incrementSlotProps = computed(() => ({
      click: onClickUp
    }));
    const decrementSlotProps = computed(() => ({
      click: onClickDown
    }));
    function toggleUpDown() {
      let increment = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
      if (model.value == null) {
        model.value = 0;
        return;
      }
      const decimals = Math.max(modelDecimals.value, stepDecimals.value);
      if (increment) {
        if (canIncrease.value) model.value = +(model.value + props.step).toFixed(decimals);
      } else {
        if (canDecrease.value) model.value = +(model.value - props.step).toFixed(decimals);
      }
    }
    function onClickUp() {
      toggleUpDown();
    }
    function onClickDown() {
      toggleUpDown(false);
    }
    function onKeydown(e) {
      if (['Enter', 'ArrowLeft', 'ArrowRight', 'Backspace'].includes(e.key) || e.ctrlKey) return;
      if (['ArrowDown'].includes(e.key)) {
        e.preventDefault();
        toggleUpDown(false);
        return;
      }
      if (['ArrowUp'].includes(e.key)) {
        e.preventDefault();
        toggleUpDown();
        return;
      }

      // Only numbers, +, - & . are allowed
      if (!/^[0-9\-+.]+$/.test(e.key)) {
        e.preventDefault();
      }
    }
    function onInput(e) {
      const el = e.target;
      model.value = el.value ? +el.value : undefined;
    }
    useRender(() => {
      const fieldProps = filterFieldProps(props);
      const [rootAttrs, inputAttrs] = filterInputAttrs(attrs);
      const {
        modelValue: _,
        ...inputProps
      } = VInput.filterProps(props);
      function controlNode() {
        const defaultHeight = controlVariant.value === 'stacked' ? 'auto' : '100%';
        return _createVNode("div", {
          "class": "v-number-input__control"
        }, [!slots.decrement ? _createVNode(VBtn, {
          "disabled": !canDecrease.value,
          "flat": true,
          "key": "decrement-btn",
          "height": defaultHeight,
          "name": "decrement-btn",
          "icon": "$expand",
          "size": "small",
          "onClick": onClickDown
        }, null) : _createVNode(VDefaultsProvider, {
          "key": "decrement-defaults",
          "defaults": {
            VBtn: {
              disabled: !canDecrease.value,
              flat: true,
              height: defaultHeight,
              size: 'small',
              icon: '$expand'
            }
          }
        }, {
          default: () => [slots.decrement(decrementSlotProps.value)]
        }), _createVNode(VDivider, {
          "vertical": controlVariant.value !== 'stacked'
        }, null), !slots.increment ? _createVNode(VBtn, {
          "disabled": !canIncrease.value,
          "flat": true,
          "key": "increment-btn",
          "height": defaultHeight,
          "name": "increment-btn",
          "icon": "$collapse",
          "onClick": onClickUp,
          "size": "small"
        }, null) : _createVNode(VDefaultsProvider, {
          "key": "increment-defaults",
          "defaults": {
            VBtn: {
              disabled: !canIncrease.value,
              flat: true,
              height: defaultHeight,
              size: 'small',
              icon: '$collapse'
            }
          }
        }, {
          default: () => [slots.increment(incrementSlotProps.value)]
        })]);
      }
      function dividerNode() {
        return !props.hideInput && !props.inset ? _createVNode(VDivider, {
          "vertical": true
        }, null) : undefined;
      }
      return _createVNode(VInput, _mergeProps({
        "class": ['v-number-input', {
          'v-number-input--default': controlVariant.value === 'default',
          'v-number-input--hide-input': props.hideInput,
          'v-number-input--inset': props.inset,
          'v-number-input--reverse': props.reverse,
          'v-number-input--split': controlVariant.value === 'split',
          'v-number-input--stacked': controlVariant.value === 'stacked'
        }, props.class]
      }, rootAttrs, inputProps, {
        "focused": isFocused.value,
        "style": props.style
      }), {
        ...slots,
        default: () => _createVNode(VField, _mergeProps(fieldProps, {
          "active": true,
          "focused": isFocused.value
        }), {
          ...slots,
          default: _ref2 => {
            let {
              props: {
                class: fieldClass,
                ...slotProps
              }
            } = _ref2;
            return _createVNode("input", _mergeProps({
              "ref": inputRef,
              "type": "text",
              "value": model.value,
              "onInput": onInput,
              "onKeydown": onKeydown,
              "class": fieldClass,
              "onFocus": onFocus,
              "onBlur": blur
            }, inputAttrs), null);
          },
          'append-inner': controlVariant.value === 'split' ? () => _createVNode("div", {
            "class": "v-number-input__control"
          }, [_createVNode(VDivider, {
            "vertical": true
          }, null), _createVNode(VBtn, {
            "flat": true,
            "height": "100%",
            "icon": "$plus",
            "tile": true,
            "onClick": onClickUp
          }, null)]) : !props.reverse ? () => _createVNode(_Fragment, null, [dividerNode(), controlNode()]) : undefined,
          'prepend-inner': controlVariant.value === 'split' ? () => _createVNode("div", {
            "class": "v-number-input__control"
          }, [_createVNode(VBtn, {
            "flat": true,
            "height": "100%",
            "icon": "$minus",
            "tile": true,
            "onClick": onClickDown
          }, null), _createVNode(VDivider, {
            "vertical": true
          }, null)]) : props.reverse ? () => _createVNode(_Fragment, null, [controlNode(), dividerNode()]) : undefined
        })
      });
    });
  }
});
